server/src/test/java/io/crate/auth/HttpAuthUpstreamHandlerTest.java

```diff
@@ -346,4 +351,89 @@ public void test_user_authentication_with_jwt_token_verified_per_connection() th
         verify(jwtAuth, times(1)).authenticate(any(Credentials.class), any(ConnectionProperties.class));
     }
 
+    @Test
+    public void test_user_has_no_jwt_properties_default_is_used() throws Exception {
+        String tokenUsername = "cloud_user"; // Token payload dictates CrateDB username.
+        Role userWithoutJWTProps = userOf(
+            tokenUsername,
+            Set.of(),
+            new HashSet<>(),
+            getSecureHash("pwd"),
+            // User doesn't have JWT properties
+            null
+        );
+        Roles roles = () -> List.of(userWithoutJWTProps);
+        Authentication authentication = mock(Authentication.class);
+        AuthenticationMethod jwtAuth = mock(JWTAuthenticationMethod.class);
+        when(authentication.resolveAuthenticationType(eq(tokenUsername), any(ConnectionProperties.class)))
+            .thenReturn(jwtAuth);
+        when(jwtAuth.authenticate(any(Credentials.class),any(ConnectionProperties.class))).thenReturn(userWithoutJWTProps);
+
+        HttpAuthUpstreamHandler handler = new HttpAuthUpstreamHandler(Settings.EMPTY, authentication, roles);
+        EmbeddedChannel ch = new EmbeddedChannel(handler);
+
+        HttpRequest request = new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.POST, "/_sql");
+        request.headers().add(HttpHeaderNames.AUTHORIZATION.toString(), "Bearer " + JWT_TOKEN);
+
+        ch.writeInbound(request);
+        ch.releaseInbound();
+
+        assertThat(handler.authorized()).isTrue();
+    }
+
+
+    /**
+     * This test demonstrates possibility of the ambiguous user resolution.
+     * Token can match:
+     *  1. User with JWT properties (token.username == crateUser.jwt.username)
+     *  2. User without JWT properties (token.username = crateUser).
+     * Match by username takes precedence as a safer option (name cannot be altered but jwt properties can).
+     */
+    @Test
+    public void test_user_lookup_matches_multiple_users() throws Exception {
```

Comment:
Same here, can't we just unit test the relevant methods?

Reply:
https://github.com/crate/crate/pull/16763#discussion_r1806378528 (even more relevant here as it's exactly about lookup aspect)

Reply:
Please replace with a dedicated test to ensure the roles predicate returned by `Credentials` works as expected. Everything else covered here is already tested.